crates/telio-starcast/src/multicast_peer.rs

```diff
@@ -0,0 +1,326 @@
+
+/// The actual multicast peer containing the task which runs the wait loop for this whole component.
+pub struct MulticastPeer {
+    task: Task<State>,
+}
+
+/// Multicast peer configuration.
+pub struct Config {
+    /// Public key of the multicast virtual peer.
+    public_key: PublicKey,
+    /// Port number through which the virtual multicast peer shall communicate with WireGuar.
+    wg_port: u16,
+}
+
+/// State for the multicast peer runtime where all the "working" members are stored and operated on.
+struct State {
+    /// IP of the multicaster.
+    multicaster_ip: MulticasterIp,
+    /// Interval at which the tunnel timers will be updated.
+    wg_timer_tick_interval: Interval,
+    /// Channel for sending decapsulated packets to the multicaster and receiving packets for encapsulation from it.
+    channel: Chan<Vec<u8>>,
+    /// Local socket for sending and receiving packets from WireGuard.
+    socket: Option<UdpSocket>,
+    /// WireGuard socket address.
+    wg_sock_addr: Option<SocketAddr>,
+    /// WireGuard encryption tunnel for encapsulation/decapsulation of packets.
+    tunnel: Option<Tunn>,
+    /// Flag indicating that there are packets in the encryption tunnel internal buffer that need to be processed
+    /// before new packets can be received from the socket and decapsulated.
+    packets_present_in_tunnel: bool,
+    // TODO need to recheck if there are no issues with these buffers:
+    /// Buffer for outbound packets.
+    sending_buffer: Vec<u8>,
+    /// Buffer for inbound packets.
+    receiving_buffer: Vec<u8>,
+}
+
+impl State {
+    async fn configure(&mut self, config: Config) -> Result<u16, Error> {
+        let socket = UdpSocket::bind(SocketAddr::from(([127, 0, 0, 1], config.wg_port))).await?;
+        let actual_port = socket.local_addr()?.port();
+        self.socket = Some(socket);
+
+        self.wg_sock_addr = Some(SocketAddr::from(([127, 0, 0, 1], config.wg_port)));
+
+        self.tunnel = Some(
+            Tunn::new(
+                StaticSecret::from(SecretKey::gen().into_bytes()),
+                PublicKeyDalek::from(config.public_key.0),
+                None,
+                None,
+                0,
+                None,
+            )
+            .map_err(Error::TunnNewFail)?,
+        );
+
+        Ok(actual_port)
+    }
+
+    fn get_peer(&self) -> Result<Peer, Error> {
+        match self {
+            Self {
+                tunnel: Some(tunnel),
+                wg_sock_addr: Some(wg_sock_addr),
+                ..
+            } => {
+                Ok(Peer {
+                    public_key: PublicKey::from(tunnel.peer_static_public().as_bytes()),
+                    endpoint: Some(wg_sock_addr.to_owned()),
+                    // Include all local non-routable ipv4 multicast addresses and single ipv6 address for just MDNS for now
+                    allowed_ips: vec![
+                        IpNetwork::V4(Ipv4Network::new(Ipv4Addr::new(224, 0, 0, 0), 4)?),
```

Comment:
This should also be a passed in argument, it will allow alligning with transport component

Reply:
Done